### DISCUSSION TITLE
Need help: Apply-like operations with Dataframes

### DISCUSSION THREAD

**Discussion by cgutierrezpa** [opensource9ja/danfojs#153] [Q&A] [answered]
First of all, thank you for anyone willing to help me, and apologies for the simple question and if it's not written in the right place, but I'm having a hard time trying to figure out the answer.

The thing is, imagine the following Dataframe:
```
let data = [["R", 100], ["L", 200], ["R", "NULL"]]
let df = new dfd.DataFrame(data, {columns:["side", "value"]})

df.print()
```
![image](https://user-images.githubusercontent.com/20757456/111973758-1f382980-8aff-11eb-8c84-67e51e462a72.png)

What I want is to add a new column to that DF based on operations performed with one or several other columns, in a Pandas-like fashion. For example, imagine that I want to create a new column **result** that takes the following rule:

```
if (side === "R" && value !== "NULL" && value > 0) {
    result = value;
} else {
    result = 0;
}
```

The idea is to obtain the following DF:

![image](https://user-images.githubusercontent.com/20757456/111974630-e482c100-8aff-11eb-84cb-a522e9388988.png)

In Pandas, this could be easily done by:
```
fn = lambda row: return row.value if row.side == "R" and row.value != "NULL" and row.value > 0 else 0
df["result"] = df.apply(fn, axis=1)
```

However, as Danfo does not support JS functions when specifying the axis in `df.apply`, I find myself trying to implement weird workarounds that don't end anywhere

Is there a way in which I could implement this using Danfo? I have to perform several operations like this one to create several columns in my Dataframe, each column having its own rules and depending on other columns (can be one or more columns), and I wish to not have to traverse the whole df row by row with a classic for loop, ending up in not getting full advantage of Danfo

**Comment by risenW** [comment]
@cgutierrezpa That's a really complicated function, and most of the parts needed to compute it properly are currently missing. 
For instance direct string comparison like pandas `df["string_col"] == "some value"` is not currently implemented. 
Also, we will need to support dataframe selection by boolean mask, and this is not currently implemented as well. 

The only way to implement this for now is hacky and you will have to write your own functions that can operate on dataframe values directly. 

This is an important function, and we'll love to add it.

**Comment by FrozenFalcon** [comment]
I believe this is another example for such operations in Pandas: https://pandas.pydata.org/docs/getting_started/intro_tutorials/05_add_columns.html

Would be great if support for this gets implemented. The current workaround is to build some kind of loop over rows but that gets ugly very quickly.

**Comment by risenW** [comment, marked as answer]
I’m case anyone comes looking. You can now use a custom JS function with apply.
